Cheat detection

trberg opened the pull request:
- [x] Fixes #xxxx (Link the github issue if applicable)
- [x] Adds in a cheat_detection module that scans an evaluation queue for users who may be collaborating to evade submission limits.
- [x] Did you run your tests locally (instructions [here](https://github.com/Sage-Bionetworks/challengeutils/blob/master/CONTRIBUTING.md))?
- [x] Did you add a good description about your changes or additions?

vpchung commented on challengeutils/__main__.py:
```diff
@@ -383,6 +384,22 @@ def command_add_mod_flair(syn, args):
+    >>> challengeutils cheat-detection <evaluation_id>
+    """
+
+    if type(args.submission_status) == str:
```
It's more preferable in Python to do duck typing rather than actually get the type.  So a better alternative would be:
```suggestion
    if isinstance(args.submission_status, str):
```

vpchung commented on tests/test_cheatdetection.py:
```diff
@@ -0,0 +1,47 @@
+
+    def test_evaluation_id(self):
+        "Test that the evaluation id is valid"
+        assert self.cheat_detection_obj.evaluation=="123456"
```
Nit: add whitespace around `==`

trberg replied:
It works with Awesome Challenge now. It wasn't able to handle challenges with one submission, now it can.

vpchung reviewed (changes requested):
This is awesome, thanks @trberg !  I just tested with two queues (one for FINRISK, one for a dummy challenge).  It worked great with FINRISK:

```sh
$ challengeutils cheat-detection 9615015 -s SCORED
Finding users in cahoots...
Calculating filename similarities...
Calculating submission cooccurrences...
No linked users found
```

But broke with [Awesome Challenge](https://www.synapse.org/#!Synapse:syn28590455/challenge/).  I think maybe because the Evaluation queue does not have any rounds?

```sh
$ challengeutils cheat-detection 9615294
Finding users in cahoots...
Calculating filename similarities...
Traceback (most recent call last):
  File "/Users/vchung/opt/miniconda3/envs/client/bin/challengeutils", line 8, in <module>
    sys.exit(main())
  File "/Users/vchung/Desktop/challengeutils/challengeutils/__main__.py", line 863, in main
    args.func(syn, args)
  File "/Users/vchung/Desktop/challengeutils/challengeutils/__main__.py", line 400, in run_cheat_detection
    cd.cheat_detection()
  File "/Users/vchung/Desktop/challengeutils/challengeutils/cheat_detection.py", line 514, in cheat_detection
    self.filename_similarity()
  File "/Users/vchung/Desktop/challengeutils/challengeutils/cheat_detection.py", line 255, in filename_similarity
    submitted_files['similarity'] = submitted_files.apply(lambda row: j.jaro_similarity(row["filename_x"], row["filename_y"]), axis=1)
  File "/Users/vchung/opt/miniconda3/envs/client/lib/python3.9/site-packages/pandas/core/frame.py", line 3970, in __setitem__
    self._set_item_frame_value(key, value)
  File "/Users/vchung/opt/miniconda3/envs/client/lib/python3.9/site-packages/pandas/core/frame.py", line 4125, in _set_item_frame_value
    raise ValueError(
ValueError: Cannot set a DataFrame with multiple columns to the single column similarity
```

vpchung reviewed (changes requested):
Thanks for the quick turnaround, @trberg! The previous error is no longer encountered:

```sh
$ challengeutils cheat-detection 9615294
Finding users in cahoots...
Calculating filename similarities...
1 valid submissions - Not enough accepted submissions to analyze queue
```

That being said, I added a couple more submissions to the Awesome Challenge queue and am encountering the error again:

```sh
$ challengeutils cheat-detection 9615294 -s RECEIVED
Finding users in cahoots...
Calculating filename similarities...
Traceback (most recent call last):
  File "/Users/vchung/opt/miniconda3/envs/client/bin/challengeutils", line 8, in <module>
    sys.exit(main())
  File "/Users/vchung/Desktop/challengeutils/challengeutils/__main__.py", line 875, in main
    args.func(syn, args)
  File "/Users/vchung/Desktop/challengeutils/challengeutils/__main__.py", line 403, in run_cheat_detection
    cheat_detect.cheat_detection()
  File "/Users/vchung/Desktop/challengeutils/challengeutils/cheat_detection.py", line 679, in cheat_detection
    self.filename_similarity()
  File "/Users/vchung/Desktop/challengeutils/challengeutils/cheat_detection.py", line 289, in filename_similarity
    submitted_files["similarity"] = submitted_files.apply(
  File "/Users/vchung/opt/miniconda3/envs/client/lib/python3.9/site-packages/pandas/core/frame.py", line 3970, in __setitem__
    self._set_item_frame_value(key, value)
  File "/Users/vchung/opt/miniconda3/envs/client/lib/python3.9/site-packages/pandas/core/frame.py", line 4125, in _set_item_frame_value
    raise ValueError(
ValueError: Cannot set a DataFrame with multiple columns to the single column similarity
```

trberg commented:
> Thanks for the quick turnaround, @trberg! The previous error is no longer encountered:
> 
> ```shell
> $ challengeutils cheat-detection 9615294
> Finding users in cahoots...
> Calculating filename similarities...
> 1 valid submissions - Not enough accepted submissions to analyze queue
> ```
> 
> That being said, I added a couple more submissions to the Awesome Challenge queue and am encountering the error again:
> 
> ```shell
> $ challengeutils cheat-detection 9615294 -s RECEIVED
> Finding users in cahoots...
> Calculating filename similarities...
> Traceback (most recent call last):
>   File "/Users/vchung/opt/miniconda3/envs/client/bin/challengeutils", line 8, in <module>
>     sys.exit(main())
>   File "/Users/vchung/Desktop/challengeutils/challengeutils/__main__.py", line 875, in main
>     args.func(syn, args)
>   File "/Users/vchung/Desktop/challengeutils/challengeutils/__main__.py", line 403, in run_cheat_detection
>     cheat_detect.cheat_detection()
>   File "/Users/vchung/Desktop/challengeutils/challengeutils/cheat_detection.py", line 679, in cheat_detection
>     self.filename_similarity()
>   File "/Users/vchung/Desktop/challengeutils/challengeutils/cheat_detection.py", line 289, in filename_similarity
>     submitted_files["similarity"] = submitted_files.apply(
>   File "/Users/vchung/opt/miniconda3/envs/client/lib/python3.9/site-packages/pandas/core/frame.py", line 3970, in __setitem__
>     self._set_item_frame_value(key, value)
>   File "/Users/vchung/opt/miniconda3/envs/client/lib/python3.9/site-packages/pandas/core/frame.py", line 4125, in _set_item_frame_value
>     raise ValueError(
> ValueError: Cannot set a DataFrame with multiple columns to the single column similarity
> ```

I've added an exception handlers to account for multiple submissions but only one user.

vpchung reviewed (approved):
LGTM 🚀
